posts/intake_esm_dask.md

```diff
@@ -0,0 +1,197 @@
+
+```python
+import dask
+
+# Use dask jobqueue
+from dask_jobqueue import PBSCluster
+
+# Import a client
+from dask.distributed import Client
+
+# Setup your PBSCluster
+cluster = PBSCluster(
+    cores=2, # The number of cores you want
+    memory='256 GB', # Amount of memory
+    processes=1, # How many processes
+    queue='casper', # The type of queue to utilize (/glade/u/apps/dav/opt/usr/bin/execcasper)
+    local_directory='$TMPDIR', # Use your local directory
+    resource_spec='select=1:ncpus=2:mem=256GB', # Specify resources
+    project='project_id', # Input your project ID here
+    walltime='02:00:00', # Amount of wall time
+    interface='ib0', # Interface to use
+)
+# Scale up
+cluster.scale(4)
+
+# Change your url to the dask dashboard so you can see it
+dask.config.set({'distributed.dashboard.link':'https://jupyterhub.hpc.ucar.edu/stable/user/{USER}/proxy/{port}/status'})
+
+# Setup your client
+client = Client(cluster)
+```
+
+## Read the CESM-LE data
+
+We will use [`intake-esm`](https://intake-esm.readthedocs.io/en/latest/), which is a data catalog tool.
+It enables querying a database for the files we want, then loading those directly as an `xarray.Dataset`.
+
+First step is to set the "collection" for the CESM-LE, which depends on a json file conforming to the [ESM Catalog Specification](https://github.com/NCAR/esm-collection-spec).
+
+```python
+catalog_file = '/glade/u/home/kristenk/TTE_CESM-LE/krill-cesm-le/notebooks/data/glade-cesm1-le.json'
+variables = ['TEMP', 'diatC']
+
+experiments = ['20C', 'RCP85']
+stream = 'pop.h'
+
+col = intake.open_esm_datastore(catalog_file, sep=',')
+col
+```
+
+Now we search the collection for the ensemble members (unique `member_id`'s) that have a chlorophyll field (`diatChl`). This is necessary because the ocean biogeochemistry was corrupted in some members and the data were deleted.
+
+In this cell, `member_id` is a list of the ensemble members we want to operate on.
+
+```python
+col_sub = col.search(experiment=['20C'],
+                     stream='pop.h',
+                     variable=['diatChl'])
+
+member_id = list(col_sub.df.member_id.unique())
+print(member_id)
+```
+
+### Apply the search for data
+
+Specify a list of variables and perform a search. Under the hood, the `search` functionality uses [`pandas`](https://pandas.pydata.org/) data frames. We can view that frame here using the `.df` syntax.
+
+```python
+col_sub = col.search(
+    experiment=experiments,
+    stream=stream,
+    variable=variables,
+    member_id=member_id,
+    )
+
+print(col_sub)
+
+col_sub.df.head()
+```
+
+Use `.to_dataset_dict` to read in the datasets as a dictionary, specifying the chunk size. This is also a point where it is recommended you take a look at the `chunksize` for the dask array, ensuring that it is a reasonable number (~100-200 mb). Test it out on one of the variables (ex. `TEMP`)
+
+```python
+dsets = col_sub.to_dataset_dict(cdf_kwargs={'chunks': {'time':5}, 'decode_times': False})
```

Comment:
You only need the top 100m so chunk in `z_t` also. I bet this reduces memory a lot. Xarray+dask will read entire files, and then subset to 100m.